File: algo/101. Symmetric Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    var helper = function(left, right) {
        if (left===null && right!==null) {
            return false;
        } else if (left!==null && right===null) {
            return false;
        } else if (left===null && right===null) {
            return true;
        } else {
            if (left.val != right.val) {
                return false;
            } else {
                return helper(left.left, right.right) && helper(left.right, right.left);
            }
        }        
    };
    
    if (root == null) {
        return true;
    }
    
    return helper(root.left, root.right);
};

File: round3/101. Symmetric Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        return Helper(root.left, root.right);
    }
    
    private bool Helper(TreeNode left, TreeNode right) {
        if (left == null && right == null) return true;
        if ((left == null && right != null) || (left != null && right == null) || (left.val != right.val)) return false;
        
        return Helper(left.left, right.right) && Helper(left.right, right.left);
    }
}